pipeline/detection.py

```diff
@@ -98,20 +201,422 @@ def run(self, *args, **kwargs):
+
+            # Get the PSF
+            psf = self._run_psfex( tempnamebase, image.id, do_not_cleanup=True )
+            psf.image = image
+        else:
+            psf = self.pars.psf
+            if psf is not None:
+                psfpath, psfxmlpath = self.psf.get_fullpath( download=download, always_verify_md5=True )
+            else:
+                psfpath = None
+                psfxmlpath = None
+
+        if self.pars.aperunit == 'fwhm':
+            if psf is None:
+                raise RuntimeError( "No psf passed to extract_sources_sextractor, so apertures can "
+                                    "not be based on the FWHM." )
+            else:
+                apers *= psf.fwhm_pixels
+
+        # Now that we have a psf, run sextractor (maybe a second time)
+        # to get the actual measurements.
+        sources = self._run_sextractor_once( image, apers=apers, psffile=psfpath, tempname=tempnamebase )
+
+        snr = sources.apfluxadu()[0] / sources.apfluxadu()[1]
+        if snr.min() > self.pars.threshold:
+            _logger.warning( "SExtractor may not have detected everything down to your threshold." )
+        w = np.where( snr >= self.pars.threshold )
+        sources.data = sources.data[w]
+        sources.num_sources = len( sources.data )
+
+        # Clean up the temporary files created (that weren't already cleaned up by _run_sextractor_once)
+        sourcepath.unlink( missing_ok=True )
+        if self.pars.psf is None:
+            if psfpath is not None: psfpath.unlink( missing_ok=True )
+            if psfxmlpath is not None: psfxmlpath.unlink( missing_ok=True )
+
+        return sources, psf
+
+    def _run_sextractor_once( self, image, apers=[5, ], psffile=None, tempname=None, do_not_cleanup=False ):
+        """Extract a SourceList from a FITS image using SExtractor.
+
+        This function should not be called from outside this class.
+
+        Parameters
+        ----------
+          image: Image
+            The Image object from which to extract.  This routine will
+            use all of image, weight, and flags data.
+
+          apers: list of float
+            Aperture radii in pixels in which to do aperture photometry.
+
+          psffile: Path or str, or None
+            File that has the PSF to use for PSF photometry.  If None,
+            won't do psf photometry.
+
+          tempname: str
+            A filename base for where the catalog will be written.  The
+            source file will be written to
+            "{FileOnDiskMixin.temp_path}/{tempname}.sources.fits".  (A
+            temporary image, weight, and mask file will be written with
+            the same name base to the same directory, but will be
+            deleted by this routine unless do_not_cleanup is False.)  It
+            is the responsibility of the calling routine to delete this
+            temporary file.
```

Comment:
It is a little bit confusing. Can you emphasize that there's a single file that must be written out and deleted by external function, and besides that there are also these additional temporary files that may or may not be deleted by this function.